### ISSUE TITLE
Improve e2e test infrastructure

### ISSUE THREAD

**Issue by kyessenov** [istio/istio#1421]
@kyessenov commented on [Thu Apr 13 2017](https://github.com/istio/old_pilot_repo/issues/533)

The integration test framework has served well but we're at the limits of what it can do. Here is the call to identify any gaps in the e2e test coverage so that we can set the goals towards building a more robust test framework. Feel free to edit this issue and add more items or mark the completed items.

Framework:
- [x] common deploy, setup, run, log checking, teardown workflow for tests
- [ ] more reliable config propagation than waiting a few seconds
- [ ] ability to stress any feature by increasing load, repeat count
- [x] test matrix to exercise feature combinations: TCP for auth, HTTP ingress for non-auth, route-splitting with auth, auth with http2 auto-upgrade, etc
- [x] stabilize flaky mixer log checking due to mixer dependency, e.g. using a fake mixer
- [x] full log aggregation on failures
- [x] reduce log noise in debug mode
- [x] ability to run subsets of tests. Useful for debug/development and if test needs to be run in parallel across multiple clusters.
- [ ] collect core dumps from envoy crashes
- [ ] make liveness probes optional to avoid crash loops

Proxy:
- [x] verify proxy with basic fixed configurations
- [x] verify traffic redirection (i.e. iptables)
- [ ] regression stress tests for envoy hot restart protocol 

Reachability:
- [x] assert two proxies on self-loops ("a" to "a")
- [x] add more pods to exercise route tables for `0.0.0.0:port`
- [ ] assert correct ports are reached
- [x] grpc and http2 ports in the test app
- [ ] validate that `NodePort` and `LoadBalancer` services work

Ingress:
- [x] verify multiple paths in the ingress
- [x] plaintext ingress
- [x] TLS in ingress
- [x] route rules applied to ingress services
- [x] missing paths should return 404

Egress:
- [x] HTTP single-port `ExternalName` service reachability
- [x] HTTPS single-port `ExternalName` service reachability 
- [x] route rules applied to egress services

Auth:
- [x] negative test to ensure protocol is upgraded
- [ ] attempt to connect to a pod that doesn't belong to a service as if it belongs to a service

Routing:
- [x] source service based routing
- [x] exact header based routing
- [x] regex header based routing
- [x] prefix header based routing
- [x] multi-condition based routing
- [x] fault injection: delay, abort
- [x] weighted route
- [x] rule precedence
- [ ] http timeout policy
- [ ] http retry policy 

Destination policies:
- [ ] per service version policy
- [ ] circuit breaking policy
- [ ] load balancing policy

---

@rshriram commented on [Tue Apr 18 2017](https://github.com/istio/old_pilot_repo/issues/533#issuecomment-294995619)

Some of the routing tests (circuit breakers, retries, timeouts) will be difficult to test in our framework reliably. E.g., a circuit breaker might need in excess of 100s of req/s in order to be tripped. An outlier detection configuration would require one or more hosts to go down in a particular order in order to be activated. 

It would be equivalent to testing Envoy's functionality rather than our configuration, won't it?

We could test these configurations with *.golden files.. And we already do for several of them.

---

@kyessenov commented on [Mon Aug 14 2017](https://github.com/istio/old_pilot_repo/issues/533#issuecomment-322378144)

this can be further improved

**Comment by kyessenov** [comment]
cc @nmittler for the ancient issue tracking integration tests in pilot.

**Comment by nmittler** [comment]
@kyessenov excellent! Ancient bugs are the best bugs!

**Comment by nmittler** [comment]
Closing as a duplicate of https://github.com/istio/istio/issues/4545
